nexus/db-model/src/instance.rs

```diff
@@ -220,6 +229,206 @@ impl InstanceRuntimeState {
             dst_propolis_id: None,
             migration_id: None,
             gen: Generation::new(),
+            time_last_auto_restarted: None,
         }
     }
 }
+
+/// Configuration for automatic instance restarts.
+#[derive(
+    Clone,
+    Debug,
+    Default,
+    AsChangeset,
+    Selectable,
+    Insertable,
+    Queryable,
+    Serialize,
+    Deserialize,
+)]
+#[diesel(table_name = instance)]
+pub struct InstanceAutoRestart {
+    /// The auto-restart policy for this instance.
+    ///
+    /// This indicates whether the instance should be automatically restarted by
+    /// the control plane on failure. If this is `NULL`, no auto-restart policy
+    /// has been configured for this instance by the user.
+    #[diesel(column_name = auto_restart_policy)]
+    #[serde(default)]
+    pub policy: Option<InstanceAutoRestartPolicy>,
+    /// The cooldown period that must elapse between automatic restarts of this
+    /// instance.
+    ///
+    /// If this is `NULL`, no explicit cooldown period has been configured for
+    /// this instance, and the default cooldown period should be used instead.
+    #[diesel(column_name = auto_restart_cooldown)]
+    #[serde(default, with = "optional_time_delta")]
+    pub cooldown: Option<TimeDelta>,
+}
+
+impl InstanceAutoRestart {
+    /// The default cooldown used when an instance has no overridden cooldown.
+    pub const DEFAULT_COOLDOWN: TimeDelta = match TimeDelta::try_hours(1) {
+        Some(delta) => delta,
+        None => unreachable!(), // 1 hour should be representable...
+    };
+
+    /// The default policy used when an instance does not override the
+    /// reincarnation policy.
+    pub const DEFAULT_POLICY: InstanceAutoRestartPolicy =
+        InstanceAutoRestartPolicy::Never;
```

Comment:
I also think that automatically restarting instances by default is probably the right thing, but there didn't seem to be consensus whilst I was working on the change. Therefore, I thought the right approach for now was to decouple implementing the mechanism from deciding what the right default was. This is why I intentionally left the policy field on individual instances `NULL` by default, rather than storing a default value in the database, to leave them "floating", so that we could just change this constant when we came to a decision about what the right default behavior was.

Reply:
We discussed this in the huddle today. My vote is also on automatically restarting failed instances by default. We can implement any user-configurable opt-out mechanism in a future PR.

Reply:
Changed the default to `BestEffort` in https://github.com/oxidecomputer/omicron/pull/6503/commits/467039b9628974d0a246d59da1e1447bb009d862. 

In case of emergency, https://github.com/oxidecomputer/omicron/pull/6503/commits/1485f728b0bdd4a36eff8ae4863e6ee792dd1fde adds a flag to the Nexus config file to explicitly disable all instance reincarnation, so that we can turn this off in case something goes horribly wrong. I doubt we'd have to use it, but it seemed like a good thing to have...